### PR TITLE
ci: use incremental build in ci

### PR DESCRIPTION
**Pull request by kt3k**
This PR enables the incremental build of rust code in CI by the following changes:

- Enables it by env var `CARGO_INCREMENTAL=1`. (`release` profile has `incremental = false` setting by default.)
- Uses `matrix.kind` in part of cache key because `test_release` and `test_debug` can't share the cache.
- Restores each source file's timestamp based on git log. ( `./tools/restore_mtime.py` )
  - git [doesn't store timestamp of files](https://git.wiki.kernel.org/index.php/Git_FAQ#Why_isn.27t_Git_preserving_modification_time_on_files.3F), and cargo triggers rebuild of source file based on timestamps. So we need to preserve timestamp of each file somehow. I found [this script](https://github.com/MestreLion/git-tools/blob/957810b10efa77d6ca4a3aa4f76883963621f31a/git-restore-mtime) (this script restores the timestamp of files based on commit log), and it seems working.

I tested this setting in my detached fork. windows job passed in 11 minutes https://github.com/kt3k/deno_ci_test/runs/2285441107

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -58,7 +58,7 @@ jobs:
           # Use depth > 1, because sometimes we need to rebuild main and if
           # other commits have landed it will become impossible to rebuild if
           # the checkout is too shallow.
-          fetch-depth: 5
+          fetch-depth: '0'
```
Comment (bartlomieju) [inline .github/workflows/ci.yml:61]: I believe this comment still stands - maybe we shouldn't go with depth 5 but 1 or 2 is highly desired

**Comment by piscisaureus** [comment]
I think you should also do this for submodules. Currently only files from the "root" git repo get treated.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (kt3k) [reply .github/workflows/ci.yml:61]: `0` means disabling shallow clone i.e. fetching all history. (this is needed for updating the modified timestamps) ref: https://github.com/actions/checkout#checkout-v2

So I'll update the comment.

**Comment by piscisaureus** [comment]
We can think of ways to make this even better, but let's land it as-is first and enjoy the 2x reduction in CI build time immediately.

**Comment by piscisaureus** [comment]
@bartlomieju You have the final say.

**Review by bartlomieju — APPROVED** [review]
> We can think of ways to make this even better, but let's land it as-is first and enjoy the 2x reduction in CI build time immediately.

Let's go 🚀

**Review by ry — APPROVED** [review]
LGTM - thank you Yoshiya!

**Pull request merged**